core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -66,6 +68,71 @@ class ReplicaManagerTest {
     metrics.close()
   }
 
+  @Test
+  def testAlterReplicaLogDirs() {
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, logDirCount = 3)
+    val config = KafkaConfig.fromProps(props)
+    val mockLogMgr = TestUtils.createLogManager(config.logDirs.map(new File(_)))
+    val workerThread = Thread.currentThread()
+
+    val replicaManager = new ReplicaManager(config, metrics, time, kafkaZkClient, new MockScheduler(time), mockLogMgr,
+      new AtomicBoolean(false), QuotaFactory.instantiate(config, metrics, time, ""), new BrokerTopicStats,
+      new MetadataCache(config.brokerId), new LogDirFailureChannel(config.logDirs.size)) {
+
+      override def getReplica(topicPartition: TopicPartition, replicaId: Int): Option[Replica] = {
+        val replica = super.getReplica(topicPartition, replicaId)
+
+        // Wait for ReplicaAlterLogDirsThread to move the replica. Do not block the ReplicaAlterLogDirsThread
+        if (!replicaAlterLogDirsManager.fetcherThreadMap.values.toSet.contains(Thread.currentThread())) {
```

Comment:
getReplica() is called in many places. So, it's not very clear which caller we try to block. Could you add some comments? Specifically, replicaManager.alterReplicaLogDirs() calls getReplica(), but it seems we don't want to block it since we want the 2nd alterReplicaLogDirs() call to happen when the first log dir movement hasn't been completed?

Reply:
The code currently only blocks the replicaManager.alterReplicaLogDirs() when there is ongoing replica movement. If there is no ongoing replica movement, the condition in `waitUntilTrue()` will be satisfied immediately. Thus the 1st replicaManager.alterReplicaLogDirs() should not block. Does this sound right?

Reply:
The following comment is added:

`This method blocks non-ReplicaAlterLogDirsThread when there is ongoing replica movement between log directories`

Reply:
Just want to clarify. Is the only reason for a customized getReplica() to make sure after getReplica is called in line 128, there won't be any ongoing log dir movement? If so, I am not sure if this is necessary, since the first alterReplicaLogDirs() doesn't really trigger any log dir movement?

Reply:
@junrao Sorry for late reply. No, the reason for customized getReplica() to ensure this: during the second `alterReplicaLogDirs()` at line 132, the test thread that is executing `alterReplicaLogDirs()` will block on the first `getReplica()` in this method until the ReplicaAlterLogDirsThread (created by the first alterReplicaLogDirs() at line 129) removes the future replica from the partition object.

In order to understand the motivation for this customized getReplica(), here is the scenario that causes the `NoSuchElementException` before this patch:

- `ReplicaAlterLogDirsThread` is created to move partition p1 from dir1 to dir2 after broker receives the `AlterReplicaLogDirsRequest`.
- Broker receives another `AlterReplicaLogDirsRequest` to move partition p1 from dir2 to dir3.
- `RequestHandlerThread` executes alterReplicaLogDirs(). Inside this method, it calls `getReplica(topicPartition, Request.FutureLocalReplicaId)` and find that the future replica already exists for partition p1.
- `ReplicaAlterLogDirsThread` calls `maybeReplaceCurrentWithFutureReplica` and removes the future replica from partition p1.
- `RequestHandlerThread` continues to execute alterReplicaLogDirs() and executes `futureReplica.log.get.dir.getParent != destinationDir`, which throws `NoSuchElementException`.

In order to reproduce the error, we need the future replica to be removed after `ReplicaAlterLogDirsThread` passes the check `getReplica(topicPartition, Request.FutureLocalReplicaId)` but before `ReplicaAlterLogDirsThread` calls `futureReplica.log.get.dir.getParent != destinationDir`. This can be done by the custom `getReplica()`. Does this make sense?

Reply:
@lindong28 : Hmm, where is call to getReplica(topicPartition, Request.FutureLocalReplicaId) inside alterReplicaLogDirs()? I only saw Partition.futureReplicaDirChanged() calling getReplica(Request.FutureLocalReplicaId), but that getReplica() is on Partition, not on ReplicaManager.

Reply:
@junrao Yeah you are right that there does not exist after this patch. It can can be found in the code before this patch, i.e. ReplicaManager.scala line 581 on the left side of https://github.com/apache/kafka/pull/5081/files.

Reply:
@lindong28 : Thanks, got it. It just seems that it's a bit weird that the test is only testing the code path that's already removed. So, not sure how useful it is. In the end, it seems that we just want to test that Partition.removeFutureLocalReplica() and Partition.maybeReplaceCurrentWithFutureReplica() can run in parallel. I am thinking that we could potentially design an alternative test. We customize logManager.replaceCurrentWithFutureLog() to wait on a CountDown latch. We then call Partition.maybeReplaceCurrentWithFutureReplica() in a separate thread and let it wait on the latch. We will then call Partition.removeFutureLocalReplica() in another thread. The main test thread will then count down the latch and wait for both threads to complete. What do you think?

Reply:
Sounds good. Thanks for the suggestion. I will write this test.

Reply:
@junrao I have added testMaybeReplaceCurrentWithFutureReplica() as suggested. And the previous test testAlterReplicaLogDirs() is removed now. Could you take another look? Thanks!